test/test_model.py

```diff
@@ -505,126 +544,130 @@ def test_relation_data_type_check(self):
-        with self.harness._event_context('foo_event'):
-            self.harness.update_relation_data(relation_id, 'remoteapp1',
-                                              {'secret': 'cafedeadbeef'})
-            self.harness.update_relation_data(relation_id, 'myapp',
-                                              {'local': 'data'})
-
-            self.harness.add_relation_unit(relation_id, 'remoteapp1/0')
-            self.harness.update_relation_data(relation_id, 'remoteapp1/0',
-                                              {'host': 'remoteapp1/0'})
-        self.model.relations._invalidate('db1')
-        self.resetBackendCalls()
-
-        rel_db1 = self.ensure_relation('db1')
-        self.harness.begin()
-        self.harness.set_leader(False)
-
-        local_app = self.harness.charm.app
+    def test_relation_local_app_data_readability_follower(
+            self, harness: ops.testing.Harness[ops.CharmBase], model: ops.Model):
+        relation_id = harness.add_relation('db1', 'remoteapp1')
+        with harness._event_context('foo_event'):
+            harness.update_relation_data(relation_id, 'remoteapp1',
+                                         {'secret': 'cafedeadbeef'})
+            harness.update_relation_data(relation_id, 'myapp',
+                                         {'local': 'data'})
+
+            harness.add_relation_unit(relation_id, 'remoteapp1/0')
+            harness.update_relation_data(relation_id, 'remoteapp1/0',
+                                         {'host': 'remoteapp1/0'})
+        model.relations._invalidate('db1')
+        self.resetBackendCalls(harness)
+
+        rel_db1 = self.ensure_relation(model, 'db1')
+        harness.begin()
+        harness.set_leader(False)
+
+        local_app = harness.charm.app
         # addressing the object is OK
         rel_db1.data[local_app]
         # nonleader units cannot read their local app databag
         # attempting to read it is not
-        with self.harness._event_context('foo_event'):
-            self.resetBackendCalls()
+        with harness._event_context('foo_event'):
+            self.resetBackendCalls(harness)
 
             with pytest.raises(ops.RelationDataError):
                 # 'local' is there, but still:
                 rel_db1.data[local_app]['local']
 
             # we didn't even get to relation-get
-            self.assertBackendCalls([('is_leader', )])
+            self.assertBackendCalls(harness, [('is_leader', )])
 
             # we can't see it but repr() works
             assert repr(rel_db1.data[local_app]) == '<n/a>'
-            self.assertBackendCalls([('is_leader', )])
+            self.assertBackendCalls(harness, [('is_leader', )])
 
             # as well as relation data repr() in general:
             assert isinstance(repr(rel_db1.data), str)
 
             expected_backend_calls = [
-                ('relation_get', 1, 'myapp/0', False),
+                ('relation_get', 0, 'myapp/0', False),
```

Comment:
This number is changed because previously `relation_id_db0` is always added to harness for all test cases since it's in the setup function. Now it's refactored to a fixture which is only used for a few test cases, so other test case relation id changes. Same below, there are a few occurrences.